src/plugins/intel_gpu/src/graph/shape_of.cpp

```diff
@@ -12,36 +12,33 @@
 namespace cldnn {
 GPU_DEFINE_PRIMITIVE_TYPE_ID(shape_of)
 
-layout shape_of_inst::calc_output_layout(shape_of_node const& node, kernel_impl_params const& impl_param) {
-    auto prim = impl_param.typed_desc<shape_of>();
-
-    data_types dt = data_types::i32;
-
-    if (prim->output_data_types[0])
-        dt = *prim->output_data_types[0];
-
+namespace {
+data_types get_output_data_type(const kernel_impl_params& impl_param) {
     if (impl_param.has_fused_primitives()) {
-        dt = impl_param.get_fused_output_layout().data_type;
+        return impl_param.get_fused_output_layout().data_type;
     }
 
-    cldnn::tensor out_size{static_cast<tensor::value_type>(prim->input_rank), 1, 1, 1};
+    const auto prim = impl_param.typed_desc<shape_of>();
+    return prim->output_data_types[0].value_or(data_types::i32);
+}
+}  // namespace
+
+layout shape_of_inst::calc_output_layout(shape_of_node const& node, kernel_impl_params const& impl_param) {
+    const auto prim = impl_param.typed_desc<shape_of>();
+    const auto dt = get_output_data_type(impl_param);
+    const auto rank = impl_param.get_input_layout(0).get_rank();
+    const cldnn::tensor out_size{static_cast<tensor::value_type>(rank), 1, 1, 1};
 
     return layout{dt, format::bfyx, out_size};
 }
 
 template<typename ShapeType>
 std::vector<layout> shape_of_inst::calc_output_layouts(shape_of_node const& /*node*/, const kernel_impl_params& impl_param) {
-    auto prim = impl_param.typed_desc<shape_of>();
-
-    data_types output_dt = prim->output_data_types[0].value_or(data_types::i32);
-    if (impl_param.has_fused_primitives()) {
```

Comment:
What is the reason of dropping data type deduction for fused primitives?

Reply:
It is not dropped, just moved to get_output_data_type() function - see [line 17](https://github.com/openvinotoolkit/openvino/pull/20983/files#diff-6720aa440603913c1ca4e73873fee209d6d337e3cf535c0ea14d981aa8ac9314R17) in this file.